File: Q.m
clear;

hbar = 1;
m = 1;

X = 10;      % 长度 
N = 200;    % 格点数
dx = 2*X/N; % 空间步长
En = 9;     % 能级个数

x = linspace(-X,X,N);

% 势能矩阵
v = 1/2*x.^2;
##v = x;
##I = find(v>-5 & v<5);
##v(I) = 0;
##I = find(v~=0);
##v(I) = 20;
V = spdiags(v',0,N,N); % NxN

% Hermitian矩阵
D = ones(N,1);
H = spdiags([1*D -2*D 1*D],-1:1,N,N);
H = (-(hbar^2)/(2*m))*(1/dx^2)*H;

A = H + V;

% 求特征值和特征向量
[Vec, Val] = eigs(A,En,0);

% 画图
for i = 1:En
    psi = Vec(:,i);
    E = Vec(i,i);
    subplot(3,3,i);
    [hAx,hL1,hL2] = plotyy(x,psi.^2/sum(psi.^2*dx),x,v);
    xlabel("x");
    ylabel(hAx(1),"|\Psi(x)|^2","fontsize",10);
    ylabel(hAx(2),"V(x)","fontsize",10);
    title_info = ["n=",mat2str(i)," E=" mat2str(E,3)];
    title(title_info);
    axis on;
    grid on;
end

File: 一维定态 Schrodinger 求解/Q.py
# 使用计算机来计算一维势能的定态波函数

from math import factorial
import numpy as np
import matplotlib.pyplot as plt
from scipy.special import hermite
from scipy.sparse import diags, spdiags
from scipy.sparse.linalg import eigsh

# 初始化参数
hbar = 1
m = 1
w = 1
X = 10
N = 2000
dx = 20 / N
En = 9


# 离散化空间坐标
x = np.linspace(-10, 10, N)

# 定义一维势能
# 有限深方势阱

v = 0.5 * w**2 * x**2

def exactPsi(x, n):
    A = ((m*w)/(np.pi*hbar))**(1/4) * 1/np.sqrt(2**n * factorial(n))
    xi = np.sqrt(m*w/hbar) * x
    Hn = hermite(n)
    psi = A * Hn(xi) * np.exp(-(xi**2)/2)
    return psi

def exactE(n):
    E = (n + 0.5) * hbar * w
    return E


# a = 6
# def voltageFunction(x):
#     if 0 < x < a:
#         return 0
#     else:
#         return 10**6 # 用大势能替代无穷大

# def exactPsi(x, n):
#     if not(0 < x < a):
#         return 0
#     A = np.sqrt(2/a)
#     psi = A * np.sin((n*np.pi)/a * x)
#     return psi
#
# def exactE(n):
#     E = (n**2 * np.pi**2 * hbar**2)/(2*m*a**2)
#     return E

# v = np.array([voltageFunction(x_) for x_ in x])
V = diags(v, 0)

# 构造 Hermitian 矩阵
A = np.ones(N)
D = spdiags([1 * A, -2 * A, 1 * A], [-1, 0, 1], N, N)
D = (-(hbar**2) / (2 * m)) * (1 / dx**2) * D

# 总哈密顿量
H = D + V

# 求特征值和特征向量
Val, Vec = eigsh(H, k=En, which='SA')

# 画图
plt.rcParams['figure.figsize'] = [12, 9]
plt.rcParams['figure.dpi'] = 300

for i in range(En):
    psi = Vec[:, i]
    E = Val[i]
    ax1 = plt.subplot(3, 3, i + 1)
    psi_exact = [exactPsi(x_, i)**2 for x_ in x]
    ax1.plot(x, psi_exact, 'r')
    ax1.plot(x, psi**2 / np.sum(psi**2 * dx),'--g', label=r'$|\Psi(x)|^2$')
    E_exact = exactE(i)
    ax1.set_xlabel('x')
    ax1.set_ylabel(r'$|\Psi(x)|^2$', color='g')
    ax1.tick_params(axis='y', labelcolor='g')
    ax1.grid(True)

    ax2 = ax1.twinx()
    ax2.plot(x, v, label='V(x)', color='b')
    ax2.set_ylabel('V(x)', color='b')
    ax2.tick_params(axis='y', labelcolor='b')
    
    plt.title(f'n={i} E={E:.3f} '+r'$E_{exact}$'+f'={E_exact:.3f}')

#plt.tight_layout() # 自动调整子图
plt.subplots_adjust(hspace=0.5, wspace=0.5)
plt.savefig("Q2.png")
plt.show()
